File: scripts/mpc.py
"""
This script launches a locomotion MPC scheme which solves repeatedly an 
optimal control problem based on the full dynamics model of the humanoid robot Talos. 
The contacts forces are modeled as 6D wrenches. 
"""

import numpy as np
import example_robot_data
from simple_mpc import RobotHandler, FullDynamicsProblem, KinodynamicsProblem, MPC

class Go2Parameters():
    def __init__(self, mpc_type):
        print(mpc_type)
        
        SRDF_SUBPATH = "/go2_description/srdf/go2.srdf"
        URDF_SUBPATH = "/go2_description/urdf/go2.urdf"
        modelPath = example_robot_data.getModelPath(URDF_SUBPATH)
        design_conf = dict(
            urdf_path=modelPath + URDF_SUBPATH,
            srdf_path=modelPath + SRDF_SUBPATH,
            robot_description="",
            root_name="root_joint",
            base_configuration="standing",
            controlled_joints_names=[
                "root_joint",
                "FL_hip_joint",
                "FL_thigh_joint",
                "FL_calf_joint",
                "FR_hip_joint",
                "FR_thigh_joint",
                "FR_calf_joint",
                "RL_hip_joint",
                "RL_thigh_joint",
                "RL_calf_joint",
                "RR_hip_joint",
                "RR_thigh_joint",
                "RR_calf_joint",
            ],
            end_effector_names=[
                "FL_foot",
                "FR_foot",
                "RL_foot",
                "RR_foot",
            ],
            hip_names=[
                "FL_thigh",
                "FR_thigh",
                "RL_thigh",
                "RR_thigh",
            ],
        )
        self.handler = RobotHandler()
        self.handler.initialize(design_conf)

        gravity = np.array([0, 0, -9.81])
        u0_forces = np.zeros(self.handler.getModel().nv + 6)
        for i in range(2):
            u0_forces[6 * i + 2] = -gravity[2] * self.handler.getMass() / 2

        if (mpc_type == "fulldynamics"):
            # Weight for base position and orientation
            w_basepos = [0, 0, 1, 0, 0, 0]

            # Weight for leg position (hip, thigh, ankle)
            w_legpos = [0.1, 0.1, 0.1]
           
            # Weight for base linear and angular velocity
            w_basevel = [10, 10, 10, 10, 10, 10]

            # Weight for leg velocity (hip, thigh, ankle)
            w_legvel = [1, 1, 1]

            # Concatenated state weight
            w_x = np.array(w_basepos + w_legpos * 4 + w_basevel + w_legvel * 4)

            # Weight for linear momentum regularization
            w_cent_lin = np.array([1, 1, 1])
            # Weight for angular momentum regularization
            w_cent_ang = np.array([0.1, 0.1, 1])

            # Weight for force regularization (reference is robot weight divided by nb of contacts)
            w_forces_lin = np.array([0.0002, 0.0002, 0.0002])
            
            # Weight for feet position tracking
            w_foot_tracking = 5000

            nu = self.handler.getModel().nv - 6
        
            self.problem_conf = dict(
                DT=0.01,
                w_x=np.diag(w_x),
                w_u=np.eye(nu) * 1e-4, # Weight for torque regularization
                w_cent=np.diag(np.concatenate((w_cent_lin, w_cent_ang))),
                gravity=gravity,
                force_size=3,
                w_forces=np.diag(w_forces_lin),
                w_frame=np.eye(3) * w_foot_tracking,
                umin=-self.handler.getModel().effortLimit[6:],
                umax=self.handler.getModel().effortLimit[6:],
                qmin=self.handler.getModel().lowerPositionLimit[7:],
                qmax=self.handler.getModel().upperPositionLimit[7:],
                mu=0.8,
                Lfoot=0.01,
                Wfoot=0.01,
                torque_limits=False,
                kinematics_limits=False,
                force_cone=False,
            )
        elif (mpc_type == "kinodynamics"):
            # Weight for base position and orientation
            w_basepos = [0, 0, 0, 0, 0, 0]

             # Weight for leg position (hip, thigh, ankle)
            w_legpos = [1, 1, 1]
            
            # Weight for base linear and angular velocity
            w_basevel = [10, 10, 10, 10, 10, 10]

            # Weight for leg velocity (hip, thigh, ankle)
            w_legvel = [0.1, 0.1, 0.1]
            
            # Concatenated weight for state regularization
            w_x = np.diag(np.array(w_basepos + w_legpos * 4 + w_basevel + w_legvel * 4))

            # Weight for force regularization (reference is robot weight divided by nb of contacts)
            w_forces_lin = np.array([0.001, 0.001, 0.001])

            # Concatenated weight for force and acceleration regularization
            w_u = np.diag(np.concatenate(
                (
                    w_forces_lin,
                    w_forces_lin,
                    w_forces_lin,
                    w_forces_lin,
                    np.ones(self.handler.getModel().nv - 6) * 1e-4,
                )
            ))
            
            # Weight for feet position tracking
            w_foot_tracking = 3000

            # Weight for linear momentum regularization
            w_cent_lin = np.array([0.1, 0.1, 1])

            # Weight for angular momentum regularization
            w_cent_ang = np.array([0.1, 0.1, 1])

            # Concatenated weight for momentum reg
            w_cent = np.diag(np.concatenate((w_cent_lin, w_cent_ang)))

            # Weight for derivative of linear momentum regularization
            w_centder_lin = np.ones(3) * 0.0

            # Weight for derivative of angular momentum regularization
            w_centder_ang = np.ones(3) * 0.1

            # Concatenated weight for derivative of momentum reg
            w_centder = np.diag(np.concatenate((w_centder_lin, w_centder_ang)))

            self.problem_conf = dict(
                DT=0.01,
                w_x=w_x,
                w_u=w_u,
                w_cent=w_cent,
                w_centder=w_centder,
                gravity=gravity,
                force_size=3,
                w_frame=np.eye(3) * w_foot_tracking,
                umin=-self.handler.getModel().effortLimit[6:],
                umax=self.handler.getModel().effortLimit[6:],
                qmin=self.handler.getModel().lowerPositionLimit[7:],
                qmax=self.handler.getModel().upperPositionLimit[7:],
                mu=0.8,
                Lfoot=0.01,
                Wfoot=0.01,
                kinematics_limits=False,
                force_cone=False,
            )
        else:
            print("Error: MPC type not recognized")
            exit()


class ControlBlockGo2():
    def __init__(self, mpc_type, motion, num_threads):
        print(mpc_type)
        self.param = Go2Parameters(mpc_type)
        self.motion = motion
        self.T = 40
        
        if mpc_type == "fulldynamics":
            problem = FullDynamicsProblem(self.param.handler)
        elif mpc_type == "kinodynamics":
            problem = KinodynamicsProblem(self.param.handler)
        problem.initialize(self.param.problem_conf)
        problem.createProblem(self.param.handler.getState(), self.T, 3, self.param.problem_conf["gravity"][2])
        
        if self.motion == "walk":
            self.T_fly = 30
            self.T_contact = 10
        elif self.motion == "jump":
            self.T_fly = 20
            self.T_contact = 100
        mpc_conf = dict(
            ddpIteration=1,
            support_force=-self.param.handler.getMass() * self.param.problem_conf["gravity"][2],
            TOL=1e-4,
            mu_init=1e-8,
            max_iters=1,
            num_threads=num_threads,
            swing_apex=0.2,
            T_fly=self.T_fly,
            T_contact=self.T_contact,
            T=self.T,
            dt=0.01,
        )

        self.mpc = MPC()
        self.mpc.initialize(mpc_conf, problem)
        self.nq = self.param.handler.getModel().nq
    
    def create_gait(self):
        """ Define contact sequence throughout horizon"""
        contact_phase_quadru = {
            "FL_foot": True,
            "FR_foot": True,
            "RL_foot": True,
            "RR_foot": True,
        }
        contact_phase_lift_FL = {
            "FL_foot": False,
            "FR_foot": True,
            "RL_foot": True,
            "RR_foot": False,
        }
        contact_phase_lift_FR = {
            "FL_foot": True,
            "FR_foot": False,
            "RL_foot": False,
            "RR_foot": True,
        }
        contact_phase_lift_all = {
            "FL_foot": False,
            "FR_foot": False,
            "RL_foot": False,
            "RR_foot": False,
        }
        contact_phases = []
        if self.motion == "walk":
            contact_phases = [contact_phase_quadru] * self.T_contact
            contact_phases += [contact_phase_lift_FL] * self.T_fly
            contact_phases += [contact_phase_quadru] * self.T_contact
            contact_phases += [contact_phase_lift_FR] * self.T_fly
        elif self.motion == "jump":
            contact_phases = [contact_phase_quadru] * self.T_contact
            contact_phases += [contact_phase_lift_all] * self.T_fly
            contact_phases += [contact_phase_quadru] * self.T_contact 

        self.mpc.generateCycleHorizon(contact_phases)

    def update_mpc(self, x_measured):
        self.mpc.iterate(x_measured[:self.nq],x_measured[self.nq:])

File: src/mpc.cpp
#include <pinocchio/fwd.hpp>
#include <pinocchio/parsers/srdf.hpp>
#include <pinocchio/parsers/urdf.hpp>
#include <pinocchio/algorithm/model.hpp>
#include <pinocchio/multibody/model.hpp>
#include <simple-mpc/kinodynamics.hpp>
#include <simple-mpc/fulldynamics.hpp>
#include <simple-mpc/mpc.hpp>
#include <simple-mpc/robot-handler.hpp>

class ControlBlock
{
public:
  ControlBlock(const std::string & mpc_type, const std::string motion_type, const int n_threads) {
    pinocchio::Model model;
    std::string urdf_path = EXAMPLE_ROBOT_DATA_MODEL_DIR "/go2_description/urdf/go2.urdf";
    std::string srdf_path = EXAMPLE_ROBOT_DATA_MODEL_DIR "/go2_description/srdf/go2.srdf";
    std::string base_joint_name ="root_joint";

    pinocchio::urdf::buildModel(urdf_path, pinocchio::JointModelFreeFlyer(), model);
    pinocchio::srdf::loadReferenceConfigurations(model, srdf_path, false);
    pinocchio::srdf::loadRotorParameters(model, srdf_path, false);

    simple_mpc::RobotModelHandler model_handler = 
      simple_mpc::RobotModelHandler(model, "standing", base_joint_name);
    
    pinocchio::SE3 ref_FL_foot = pinocchio::SE3::Identity();
    pinocchio::SE3 ref_FR_foot = pinocchio::SE3::Identity();
    pinocchio::SE3 ref_RL_foot = pinocchio::SE3::Identity();
    pinocchio::SE3 ref_RR_foot = pinocchio::SE3::Identity();
    ref_FL_foot.translation() = Eigen::Vector3d(0.17, 0.15, 0.0);
    ref_FR_foot.translation() = Eigen::Vector3d(0.17, -0.15, 0.0);
    ref_RL_foot.translation() = Eigen::Vector3d(-0.24, 0.15, 0.0); 
    ref_RR_foot.translation() = Eigen::Vector3d(-0.24, -0.15, 0.0);
    model_handler.addFoot("FL_foot", base_joint_name, ref_FL_foot);
    model_handler.addFoot("FR_foot", base_joint_name, ref_FR_foot);
    model_handler.addFoot("RL_foot", base_joint_name, ref_RL_foot);
    model_handler.addFoot("RR_foot", base_joint_name, ref_RR_foot);

    simple_mpc::RobotDataHandler data_handler = simple_mpc::RobotDataHandler(model_handler);

    int nu = model_handler.getModel().nv - 6;
    int nv = model_handler.getModel().nv;
    int force_dim = 3 * model_handler.getFeetNames().size();
    int nu_kino = nu + force_dim;
    int ndx = nv * 2;
    Eigen::Vector3d gravity;
    gravity << 0., 0., -9.81;

    std::shared_ptr<simple_mpc::OCPHandler> ocpPtr;

    if (mpc_type == "fulldynamics") {
      simple_mpc::FullDynamicsSettings problem_settings;

      Eigen::VectorXd w_x_vec(ndx);
      w_x_vec << 0, 0, 0, 10., 10., 0, // Base pos/ori
        1., 1., 1., 1., 1., 1.,    // FL FR leg
        1., 1., 1., 1., 1., 1.,    // RL RR leg
        10., 10., 10., 1., 1., 10.,  // Base vel
        .1, .1, .1, .1, .1, .1,     // FL FR vel
        .1, .1, .1, .1, .1, .1;    // RL RR vel 
      Eigen::VectorXd w_cent(6);
      w_cent << .1, .1, 1., 0.1, 0.1, 1;
      Eigen::VectorXd w_forces(3);
      w_forces << 0.0002, 0.0002, 0.0002;

      Eigen::VectorXd u0 = Eigen::VectorXd::Zero(nu);

      problem_settings.timestep = 0.01;
      problem_settings.w_x = Eigen::MatrixXd::Zero(ndx, ndx);
      problem_settings.w_x.diagonal() = w_x_vec;
      problem_settings.w_u = Eigen::MatrixXd::Identity(nu, nu) * 1e-4;
      problem_settings.w_cent = Eigen::MatrixXd::Zero(6, 6);
      problem_settings.w_cent.diagonal() = w_cent;
      problem_settings.gravity = gravity;
      problem_settings.force_size = 3;
      problem_settings.Kp_correction = Eigen::VectorXd::Zero(3);
      problem_settings.Kd_correction = Eigen::VectorXd::Zero(3);
      problem_settings.w_forces = Eigen::MatrixXd::Zero(3, 3);
      problem_settings.w_forces.diagonal() = w_forces;
      problem_settings.w_frame = Eigen::MatrixXd::Identity(3, 3) * 5000;
      problem_settings.umin = -model_handler.getModel().effortLimit.tail(nu);
      problem_settings.umax = model_handler.getModel().effortLimit.tail(nu);
      problem_settings.qmin = model_handler.getModel().lowerPositionLimit.tail(nu);
      problem_settings.qmax = model_handler.getModel().upperPositionLimit.tail(nu);
      problem_settings.mu = 0.8;
      problem_settings.Lfoot = 0.1;
      problem_settings.Wfoot = 0.075;
      problem_settings.torque_limits = false;
      problem_settings.kinematics_limits = false;
      problem_settings.force_cone = false;

      std::shared_ptr<simple_mpc::OCPHandler> ocpPtr = std::make_shared<simple_mpc::FullDynamicsOCP>(problem_settings, model_handler);
    }
    if (mpc_type == "kinodynamics") {
      simple_mpc::KinodynamicsSettings problem_settings;

      Eigen::VectorXd w_x_vec(ndx);
      w_x_vec << 0, 0, 0, 10., 10., 0, // Base pos/ori
        1., 1., 1., 1., 1., 1.,    // FL FR leg
        1., 1., 1., 1., 1., 1.,    // RL RR leg
        10., 10., 10., 1., 1., 10.,  // Base vel
        .1, .1, .1, .1, .1, .1,     // FL FR vel
        .1, .1, .1, .1, .1, .1;    // RL RR vel 
      
      Eigen::VectorXd w_forces(3);
      w_forces << 0.001, 0.001, 0.001;

      Eigen::VectorXd w_u_vec(nu_kino);
      w_u_vec.setZero();
      w_u_vec.tail(nv - 6) = Eigen::VectorXd::Ones(nv - 6) * 1e-4;
      w_u_vec.head(force_dim) << w_forces, w_forces, w_forces, w_forces;

      Eigen::VectorXd w_cent(6);
      w_cent << 0.1, 0.1, 1., 0.1, 0.1, 1.;

      Eigen::VectorXd w_cent_der(6);
      w_cent_der << 0., 0., 0., 0.1, 0.1, 0.1;

      problem_settings.timestep = 0.01;
      problem_settings.w_x = Eigen::MatrixXd::Zero(ndx, ndx);
      problem_settings.w_x.diagonal() = w_x_vec;
      problem_settings.w_u = Eigen::MatrixXd::Zero(nu_kino, nu_kino);
      problem_settings.w_cent = Eigen::MatrixXd::Zero(6, 6);
      problem_settings.w_cent.diagonal() = w_cent;
      problem_settings.w_centder = Eigen::MatrixXd::Zero(6, 6);
      problem_settings.w_cent.diagonal() = w_cent_der;
      problem_settings.gravity = gravity;
      problem_settings.force_size = 3;
      problem_settings.w_frame = Eigen::MatrixXd::Identity(3, 3) * 5000;
      problem_settings.qmin = model_handler.getModel().lowerPositionLimit.tail(nu);
      problem_settings.qmax = model_handler.getModel().upperPositionLimit.tail(nu);
      problem_settings.mu = 0.8;
      problem_settings.Lfoot = 0.1;
      problem_settings.Wfoot = 0.075;
      problem_settings.kinematics_limits = false;
      problem_settings.force_cone = false;

      std::shared_ptr<simple_mpc::OCPHandler> ocpPtr = std::make_shared<simple_mpc::KinodynamicsOCP>(problem_settings, model_handler);
    }
    size_t T = 100;
    ocpPtr->createProblem(model_handler.getReferenceState(), T, 3, gravity[2], false);

    simple_mpc::MPCSettings mpc_settings;
    std::size_t T_fly;
    std::size_t T_contact;
    if (motion_type == "walk") {
      T_fly = 30;
      T_contact = 5;
    }
    if (motion_type == "jump") {
      T_fly = 20;
      T_contact = 100;
    }
    mpc_settings.support_force = -gravity[2] * model_handler.getMass();
    mpc_settings.TOL = 1e-4;
    mpc_settings.mu_init = 1e-8;
    mpc_settings.max_iters = 1;
    mpc_settings.num_threads = n_threads;
    mpc_settings.T_fly = T_fly;
    mpc_settings.T_contact = T_contact;

    mpc_ = std::make_shared<simple_mpc::MPC>(mpc_settings, ocpPtr);

    std::vector<std::map<std::string, bool>> contact_states;
    std::map<std::string, bool> contact_state_quadru;
    std::map<std::string, bool> contact_phase_lift_FL_RR;
    std::map<std::string, bool> contact_phase_lift_RL_FR;
    std::map<std::string, bool> contact_phase_lift_all;

    contact_state_quadru.insert({"FL_foot", true});
    contact_state_quadru.insert({"FR_foot", true});
    contact_state_quadru.insert({"RL_foot", true});
    contact_state_quadru.insert({"RR_foot", true});

    contact_phase_lift_FL_RR.insert({"FL_foot", false});
    contact_phase_lift_FL_RR.insert({"FR_foot", true});
    contact_phase_lift_FL_RR.insert({"RL_foot", true});
    contact_phase_lift_FL_RR.insert({"RR_foot", false});

    contact_phase_lift_RL_FR.insert({"FL_foot", true});
    contact_phase_lift_RL_FR.insert({"FR_foot", false});
    contact_phase_lift_RL_FR.insert({"RL_foot", false});
    contact_phase_lift_RL_FR.insert({"RR_foot", true});

    contact_phase_lift_all.insert({"FL_foot", false});
    contact_phase_lift_all.insert({"FR_foot", false});
    contact_phase_lift_all.insert({"RL_foot", false});
    contact_phase_lift_all.insert({"RR_foot", false});
    // std::vector<std::vector<bool>> contact_states;

    if (motion_type == "walk") {
      for (std::size_t i = 0; i < T_contact; i++)
        contact_states.push_back(contact_state_quadru);
      for (std::size_t i = 0; i < T_fly; i++)
        contact_states.push_back(contact_phase_lift_FL_RR);
      for (std::size_t i = 0; i < T_contact; i++)
        contact_states.push_back(contact_state_quadru);
      for (std::size_t i = 0; i < T_fly; i++)
        contact_states.push_back(contact_phase_lift_RL_FR);
    }
    if (motion_type == "jump") {
      for (std::size_t i = 0; i < T_contact; i++)
        contact_states.push_back(contact_state_quadru);
      for (std::size_t i = 0; i < T_fly; i++)
        contact_states.push_back(contact_phase_lift_all);
      for (std::size_t i = 0; i < T_contact; i++)
        contact_states.push_back(contact_state_quadru);
    }

    mpc_->generateCycleHorizon(contact_states);
  }
  virtual ~ControlBlock() {};
  
  std::shared_ptr<simple_mpc::MPC> mpc_;
};